C#: Fix Generated ScriptProperty Error.

magian1127 opened the pull request:
Add "this." to prevent errors caused by duplicate variable names.
Try to find the default value of property getters.

加上"this.",防止和生成的变量名重名导致的错误.
尝试查找属性的默认值.

*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot/issues/66819.

neikeq commented:
I will need some example script code to understand what this fixes.

magian1127 commented:
```
using Godot;
using Godot.Collections;
using System;

public partial class new_script : Godot.RefCounted
{
	string name = "test";

    public Array<string> Skills { get; set; } = new Array<string>();

}
```

These codes generate two error reports

![image](https://user-images.githubusercontent.com/14800320/190839727-06237ba3-8b3d-4aeb-a4ed-31a053f39fc8.png)


[test4bug.zip](https://github.com/godotengine/godot/files/9590879/test4bug.zip)

magian1127 commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertiesGenerator.cs:
```diff
@@ -317,7 +317,7 @@ bool isFirstEntry
                 .Append(propertyMemberName)
                 .Append(") {\n")
                 .Append("            value = ")
-                .AppendManagedToNativeVariantExpr(propertyMemberName, propertyMarshalType)
+                .AppendManagedToNativeVariantExpr("this." + propertyMemberName, propertyMarshalType)
```
can't put it in the front. Look inside the "AppendManagedToNativeVariantExpr" to see why.

neikeq replied:
Yes, this is the way to do it right now. That being said, I'm sure we can come with a good way to optimize this in the future. Something like a wrapper struct that contains string fields to concat.

raulsntos commented:
I still think it would be preferable to always use fully qualified names (and not omitting `global::`) in generated code instead of adding usings. To clarify what this PR intends to fix is the generator for `GetGodotPropertyDefaultValues` which, currently, for an exported property like this:

```csharp
[Export]
public Array<Node> MyNodes { get; set; } = new Array<Node>();
```

Generates this code:

```csharp
internal new static System.Collections.Generic.Dictionary<StringName, object> GetGodotPropertyDefaultValues()
{
    var values = new System.Collections.Generic.Dictionary<StringName, object>(1);
    Godot.Collections.Array<Godot.Node> __MyNodes_default_value = new Array<Node>();
    values.Add(PropertyName.MyNodes, __MyNodes_default_value);
    return values;
}
```

The issue is the default value expression isn't fully qualified since it's taken as is from the source.

This PR intends to fix this by including the same usings in the generated file as the ones declared in the source, this could still conflict and I'd prefer if we didn't omit global either.

This code would still fail even with this PR:

```csharp
// using Godot; // <-- No using Godot namespace

public partial class Sprite2D : Godot.Resource {} // <- class with same name as one of the classes in Godot namespace, can be defined in a different file

public partial class MyNode : Godot.Node
{
    [Godot.Export]
    public Sprite2D Sprite { get; set; } = new Sprite2D();
}
```

---

This PR also fixes conflicts when the exported members share the same name with a local variable defined in the generated code by adding `this.`, I have no problem with this part of the PR.

magian1127 commented:
> ```cs
> // using Godot; // <-- No using Godot namespace
> 
> public partial class Sprite2D : Godot.Resource {} // <- class with same name as one of the classes in Godot namespace, can be defined in a different file
> 
> public partial class MyNode : Godot.Node
> {
>     [Godot.Export]
>     public Sprite2D Sprite { get; set; } = new Sprite2D();
> }
> ```
>
In the latest PR, this paragraph is valid.

At the beginning, I really wanted to use the fully qualified name, but after a long time of searching the roslyn document, I couldn't find a good way, so I had to do it first.


在最新的PR中,这段是有效的.
刚开始的时候,我确实想用完全限定名,但是翻了很久的roslyn文档,也没找到好的办法,就只能先这样了.

raulsntos reviewed (commented):
Seems to work well. Thanks for the contribution!

My main concern is the case where the default value is the result of an expensive operation because it will be executed on each call to `GetGodotPropertyDefaultValues` and it can be problematic if the operation has side-effects. However, this was already how it worked before this PR, I just wanted to bring attention to it.

```csharp
[Export]
public int MyProperty { get; set; } = MyExpensiveMethodWithSideEffects();
```

Also, we may want to add an example to the `Godot.SourceGenerators.Sample` project with the syntax that this PR adds support for:

```csharp
private int _myProperty = 42;

[Export]
public int MyProperty
{
	get => _myProperty;
	set => _myProperty = value;
}
```

magian1127 commented:
Generally speaking, the default value of the attribute should not change.
Users should be advised to avoid using dynamic default values.
MyExpensiveMethodWithSideEffects(), perhaps this approach could be adopted.

```C#
[Export]
public int MyProperty { get; set; } = MyExpensiveMethodWithSideEffects();
```

```C#
//generated
    private int __MyProperty = MyExpensiveMethodWithSideEffects();

    internal new static System.Collections.Generic.Dictionary<StringName, object> GetGodotPropertyDefaultValues()
    {
        ...
        string __MyPropert_default_value = __MyProperty;
        ...
    }

```

通常来说,属性默认值应该是不太会变动的.
应该建议用户,避免使用动态的默认值.
MyExpensiveMethodWithSideEffects()这种,或许可以采用这样的办法.

raulsntos reviewed (approved):
I think this looks good to merge already, @neikeq may to review this too since he self-requested a review.

magian1127 commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:
```diff
@@ -163,19 +164,83 @@ INamedTypeSymbol symbol
+                                var fieldSymbol = members
+                                    .Where(s => s.Kind == SymbolKind.Field)
+                                    .Cast<IFieldSymbol>()
+                                    .Where(s => s.Name == returnStatementSyntax?.Expression?.ToString())
```
The string check should be OK, but the `IndentifierNameSyntax` is really more reasonable

neikeq commented:
Yes, I'll review this tonight. Property default values are delicate so I want to make sure we're not doing anything undesirable.

neikeq commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:
```diff
@@ -163,19 +164,81 @@ INamedTypeSymbol symbol
+                        {
+                            if (propertyGet.ExpressionBody.Expression is IdentifierNameSyntax identifierNameSyntax)
+                            {
+                                var fieldSymbol = members
+                                    .Where(s => s.Kind == SymbolKind.Field)
+                                    .Cast<IFieldSymbol>()
+                                    .Where(s => s.Name == identifierNameSyntax.Identifier.Text)
+                                    .FirstOrDefault();
```
This is likely error-prone. I think the following is the correct way to do it, but I can't confirm as I didn't try it:

```suggestion
                                var sm = context.Compilation.GetSemanticModel(identifierNameSyntax.SyntaxTree);

                                var fieldSymbol = sm.GetSymbolInfo(identifierNameSyntax).Symbol as IFieldSymbol;
```

magian1127 replied:
Both GetSemanticModel and Linq I tested without problems

neikeq commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:
```diff
@@ -163,19 +164,81 @@ INamedTypeSymbol symbol
+                            {
+                                if (returnStatementSyntax.Expression is IdentifierNameSyntax identifierNameSyntax)
+                                {
+                                    var fieldSymbol = members
+                                        .Where(s => s.Kind == SymbolKind.Field)
+                                        .Cast<IFieldSymbol>()
+                                        .Where(s => s.Name == identifierNameSyntax.Identifier.Text)
+                                        .FirstOrDefault();
```
ditto

> This is likely error-prone. I think the following is the correct way to do it, but I can't confirm as I didn't try it:
> 
> ```suggestion
>                                 var sm = context.Compilation.GetSemanticModel(identifierNameSyntax.SyntaxTree);
> 
>                                 var fieldSymbol = sm.GetSymbolInfo(identifierNameSyntax).Symbol as IFieldSymbol;
> ```

neikeq commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:
```diff
@@ -163,19 +164,81 @@ INamedTypeSymbol symbol
                 // Fully qualify the value to avoid issues with namespaces.
                 string? value = null;
-                if (initializer != null)
+                if (propertyDeclarationSyntax?.Initializer != null)
```
I would prefer both this `if` block and the `if (string.IsNullOrWhiteSpace(value))` block below to be inside a `if (propertyDeclarationSyntax != null)` block. Seems clearer to me when reading, and the generated code may be better. I haven't checked this, but I think the current one would do unnecessary additional checks when null (`string.IsNullOrWhiteSpace(value)` and `propertyGet != null` below).

magian1127 replied:
With the Initializer, it is indeed impossible to have the latter, and here the code should be changed.

neikeq commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:
```diff
@@ -163,19 +164,81 @@ INamedTypeSymbol symbol
+                        }
+                        else
+                        {
+                            var returnStatementSyntax = propertyGet.DescendantNodes().OfType<ReturnStatementSyntax>().FirstOrDefault();
```
This doesn't make sure that the body of the getter is a single return statement, it only looks for that return statement. This can have many problems, even if the return example is just an identifier for a field. For example, a method could have more than one return statement.

IMO, the best we can support here is bodies that contain only a return statement with the identifier (as weird as it would be to write that with a block rather than an expression body). This code should make sure that's all the code does.

magian1127 replied:
only generate a single return, which should be applicable to most cases

magian1127 committed:
C#: Fix Generated ScriptProperty Error.
1. Add "this." to prevent errors caused by duplicate variable names.
2. Try to find the default value of property getters.

magian1127 commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs:
```diff
@@ -163,19 +164,81 @@ INamedTypeSymbol symbol
+                            else
+                            {//Expression not single field
+                                var sm = context.Compilation.GetSemanticModel(propertyGet.ExpressionBody.Expression.SyntaxTree);
+                                value = propertyGet.ExpressionBody.Expression.FullQualifiedSyntax(sm);
```
After thinking about it, it is better to remove it.
As mentioned above, Users should be advised to avoid using dynamic default values.

neikeq reviewed (approved):
Looks good to merge.

I think the code to get the initializer and it's value is duplicated in 3 places now, so it's a good candidate to extract to a method. But it's not necessary in order to merge.

neikeq commented:
Thanks!

magian1127 commented:
> Looks good to merge.
> 
> I think the code to get the initializer and it's value is duplicated in 3 places now, so it's a good candidate to extract to a method. But it's not necessary in order to merge.

For me, naming the method is a very distressing thing ha.
There are 4 places where this paragraph is used, and the next time you refactor the file you can consider extracting it into the method.
```
var sm = context.Compilation.GetSemanticModel(initializer.SyntaxTree);
value = initializer.Value.FullQualifiedSyntax(sm);
```
